installdir/hlstat.1.in
.TH hldig 1 "January 2004"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
hlstat \-  returns statistics on the document and word databases, much like the -s option to hldig or hlmerge.
.SH SYNOPSIS
.B hlstat
.I "[-v][-a][-c configfile][-u]"
.SH "DESCRIPTION"
hldig retrieves HTML documents using the HTTP protocol and
gathers information from these documents which can later be
used to search these documents. This program can be referred to
as the search robot. 
.SH OPTIONS
.TP
.B \-a
Use alternate work files. Tells hlstat to append
.I .work
to database files, causing a second copy of the database to be
built. This allows the original files to be used by hlsearch
during the run. 
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-u
Give a list of URLs in the document database.
.TP
.B \-v
Verbose mode. This increases the verbosity of the program.
Using more than 2 is probably only useful for debugging
purposes. The default verbose mode (using only one \fI\-v\fR) gives
a nice progress report while digging. 
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the hldig-doc package)
.B /usr/share/doc/hldig-doc/html/index.html
and the manual pages
.I hldigconfig(8)
,
.I hldig(1)
and
.I hlmerge(1)
for a detailed description of hl://Dig and its commands.

.SH AUTHOR
This manual page was written by Robert Ribnitz, 
based on the HTML documentation of hl://Dig.


ChangeLog
2018-04-04  Johnny A. Solbu <johnny@solbu.net>

	* Remove broken build test link
	* rename ht -> hl in topdir files
	* Complete rename in installdir
	* Autoreconf
	* Apparently several subscripts still use the VERSION file. Should
	fix that sometime
	* Rename mans in configure and make scripts.
	* Rename man files and their contents

2018-04-02  Johnny A. Solbu <johnny@solbu.net>

	* VERSION file is obsolete now that version is in configure.ac
	[skip ci]
	* Update pot file: copyright are assigned to «the hl://Dig
	group» [skip ci]
	* commit fb7ea686ffe3fd4aac7e823f1855fb34a3f68dd3 Author: Johnny
	A. Solbu <johnny@solbu.net> Date:   Mon Apr 2 13:31:43 2018 +0200

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* dbbench.cc:fix conversion warning This should fix the warning when `make dist` is run: dbbench.cc:269:20: warning: ISO C++ forbids converting a string
	constant to ‘char*’ [-Wpedantic]    char
	*progname = "dbbench problem...";

2018-03-28  Johnny A. Solbu <johnny@solbu.net>

	* bump version: 0.2.0
	* Add bug address to configure
	* Put version in the configure.ac file, where it belongs
	* Update build tests location
	* Rename url in the last files

2018-03-27  Johnny A. Solbu <johnny@solbu.net>

	* Rename urls in more files
	* Change MSGID_BUGS_ADDRESS, point to the issue tracker
	* Update ChangeLog.md with the new URLs
	* New project management: Rename URLs
	* Don't track compiled translations in revision controll
	* Remove unneeded preceding slash

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* change htdig.gif to logo.jpg (#56) After changes by @jamessouth, I found a few other areas I forgot to
	mention to him.

2018-03-15  jamessouth <jamesdanielsouth@gmail.com>

	* issue 56 new logo and modified html files

2018-03-15  Andy Alt <andy5995@users.noreply.github.com>

	* Changlog.md:add note about default install paths [skip ci]
	* Makefile.am:remove CVS dist-hook [skip ci]
	* CONTRIBUTING.md:add note about "skip ci" [skip ci]
	* TESTING.md:updated [skip ci] This reflects changes to the new install locations, which was a
	recent patch submitted by @Solbu
	* push regenerated build scripts and makefiles Regenerated due to previous commit
	* configure.ac:change path of www images [skip ci] When installdir/Makefile is run, it uses the value here to prefix
	urls to images in the installdir/html templates with /www/ To be accessible, a symlink can be put in the web root that points
	to that directory.

2018-03-15  Johnny A. Solbu <github@solbu.net>

	* Install things in the proper places (#90)
	  Install things in PREFIX/share/hldir, rather than /var/www

2018-03-12  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #101 from solbu/installer-fixes Installer fixes (Issue #99)
	* republish site from changes in bcdf311 This republishes the site using changes @jamessouth made in
	bcdf3110c78a308404431f3f3a24d33503d283a5 (#81)

2018-03-12  james south <jamessouth@users.noreply.github.com>

	* make sure home page validates (#100) (fixes #81) [skip ci]

2018-03-12  Johnny A. Solbu <johnny@solbu.net>

	* Add ABOUT-NLS to distributed files, indent traling backslash
	* Add missing gettext.h header

2018-03-11  andy5995 <andy400-dev@yahoo.com>

	* ABOUT-NLS:add to distribution [skip ci]
	* remove '#' (forgot to do this in a previous commit) [skip ci]
	* .travis.yml:drop '#' from hldig Freenode channel [skip ci]
	* simplecgen:refactor trim() function
	* list #hldig for the chat room, not ##hldig [skip ci]
	* .travis.yml: change failure and error notifications Only notify on_failure if there's a change Always notify on error
	* .travis.yml:change irc notification options
	* .travis.yml:change osx and Xcode version (#98)
	* .travis.yml:change osx and Xcode version
	* .travis.yml:add matrix closes #76

2018-03-11  Johnny A. Solbu <github@solbu.net>

	* Fix test dir (issue #91) (#97) / Thanks @solbu
	* Fix «make dist» Some tools where renamed at some point but the test commands using
	them where not updated
	* Update test/test_function.in with the renamed filenames

2018-03-09  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #92 from solbu/doc-changes [skip ci] Simplify translation instructions

2018-03-07  andy5995 <andy400-dev@yahoo.com>

	* translation:add ru to LINGUAS, generate gmo file (closes #68)
	* travis.yml:add osx (#76)
	* ne.po:remove empty file [skip ci]

2018-03-07  Svitlana Galianova <svitlana.galianova@gmail.com>

	* added russian translation

2018-03-07  Johnny A. Solbu <github@solbu.net>

	* Uninstall fix (#84) (closes #48)
	* Downgrade required gettext version, to match my system
	* Add Solbu to authors list

2018-03-06  andy5995 <andy400-dev@yahoo.com>

	* ChangeLog.md:note about html docs not being installed anymore

2018-03-06  Johnny A. Solbu <johnny@solbu.net>

	* Run autoreconf
	* Don't install documentaion in docdir RPM packagers prefer to control that dir using a «%doc» macro
	* Ignore .patch files To avoid committing patches by mistake, when one is also working on
	packaging

2018-01-21  andy5995 <andy400-dev@yahoo.com>

	* fix formatting [skip ci]
	* remove Slack, use IRC found some leftovers to remove (see previous commit)
	* remove Slack workspace from contact info [skip ci] IRC should be sufficient for this project.  http://webchat.freenode.net?channels=%23%23hldig&uio=d4

2018-01-15  andy5995 <andy400-dev@yahoo.com>

	* docs/README: explain how to view the html files

2018-01-11  Andy Alt <andy5995@users.noreply.github.com>

	* [#24]installdirs:update search result html, add css (#82)
	* [#24]installdirs:update search result html, add css This will copy the same style.css that the web site uses. That seems
	easier than maintaining 2 separate css files.  I added a script that runs tidy-html5, but the user must install the
	program <http://www.htacg.org/> to use it. 
	* hlsearch results html pages updated [#24] See notes in the

hltools/hlstat.cc
//
// htstat.cc
//
// htstat: A utility to give statistics on the contents of the word and doc DB.
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: htstat.cc,v 1.6 2004/05/28 13:15:25 lha Exp $
//
#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include "WordContext.h"
#include "HtURLCodec.h"
#include "HtWordList.h"
#include "HtConfiguration.h"
#include "DocumentDB.h"
#include "defaults.h"
#include "messages.h"

#include <errno.h>

#ifndef _MSC_VER                /* _WIN32 */
#include <unistd.h>
#endif

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

int verbose = 0;

void usage ();
void reportError (char *msg);

//*****************************************************************************
// int main(int ac, char **av)
//
int
main (int ac, char **av)
{
  int alt_work_area = 0;
  int url_list = 0;
  String configfile = DEFAULT_CONFIG_FILE;
  int c;
  extern char *optarg;

  while ((c = getopt (ac, av, "vc:au")) != -1)
  {
    switch (c)
    {
    case 'c':
      configfile = optarg;
      break;
    case 'v':
      verbose++;
      break;
    case 'a':
      alt_work_area++;
      break;
    case 'u':
      url_list++;
      break;
    case '?':
      usage ();
      break;
    }
  }

  HtConfiguration *config = HtConfiguration::config ();
  config->Defaults (&defaults[0]);

  if (access ((char *) configfile, R_OK) < 0)
  {
    reportError (form (_("Unable to find configuration file '%s'"),
                       configfile.get ()));
  }

  config->Read (configfile);

  //
  // Check url_part_aliases and common_url_parts for
  // errors.
  String url_part_errors = HtURLCodec::instance ()->ErrMsg ();

  if (url_part_errors.length () != 0)
    reportError (form (_("Invalid url_part_aliases or common_url_parts: %s"),
                       url_part_errors.get ()));


  // We may need these through the methods we call
  if (alt_work_area != 0)
  {
    String configValue;

    configValue = config->Find ("word_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("word_db", configValue);
    }

    configValue = config->Find ("doc_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_db", configValue);
    }

    configValue = config->Find ("doc_index");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_index", configValue);
    }

    configValue = config->Find ("doc_excerpt");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_excerpt", configValue);
    }
  }

  DocumentDB docs;
  if (docs.Read (config->Find ("doc_db"), config->Find ("doc_index"),
                 config->Find ("doc_excerpt")) == OK)
  {
    List *urls = docs.URLs ();
    cout << _("hlstat: Total documents: ") << urls->Count () << endl;
    if (url_list)
    {
      // Spit out the list of URLs too
      String *url;

      cout << _("hlstat: URLs in database: ") << endl;
      urls->Start_Get ();
      while ((url = (String *) urls->Get_Next ()))
      {
        cout << "\t" << url->get () << endl;
      }
    }

    delete urls;
    docs.Close ();
  }

  // Initialize htword
  WordContext::Initialize (*config);

  HtWordList words (*config);
  if (words.Open (config->Find ("word_db"), O_RDONLY) == OK)
  {
    cout << _("hlstat: Total words: ") << words.WordRefs ()->Count () << endl;
    cout << _("hlstat: Total unique words: ") << words.Words ()->
      Count () << endl;
    words.Close ();
  }

  return 0;
}


//*****************************************************************************
// void usage()
//   Display program usage information
//
void
usage ()
{
  Usage help;
  cout << _("usage:");
  cout << " hlstat [-v][-a][-c configfile][-u]\n";
  printf (_("This program is part of hl://Dig %s\n\n"), VERSION);
  cout << _("Options:\n");
  help.verbose ();
  help.alternate_common ();
  cout << "\t\tTells hlstat to append .work to the database files \n";
  cout << "\t\tallowing it to operate on a second set of databases.\n";
  help.config ();
  cout << "\t-u\tGive a list of URLs in the document database.\n\n";
  exit (0);
}


//*****************************************************************************
// Report an error and die
//
void
reportError (char *msg)
{
  cout << "hlstat: " << msg << "\n\n";
  exit (1);
}
